### PR TITLE
Implement Rope Mechanic

### PR DESCRIPTION
**Pull request by zahranikbakht**
To test the rope mechanic from underground, make sure player (1) and his ladder are active. interact with the ladder to climb up.
To test it from the surface, move player to the surface (y=1.4) and start the game. You should dig a hole before you can throw a rope

### CODE REVIEW

**Review by danielwiktorczyk — CHANGES_REQUESTED** [review]
>> will add more soon!

File: Assets/Scripts/Digging/Digging.cs [hunk 1]
```diff
@@ -45,13 +47,39 @@ void Update()
         var hitGameObject = closestCollider.transform.parent.gameObject;
 
         var tile = hitGameObject.GetComponent<Tile>();
-        if (tile.tileState != TileState.Normal)
+
+        if (playerMovement.IsFalling())
+        {
+            this.rope.gameObject.SetActive(false);
+            tile.isRopeIn = false;
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Digging/Digging.cs:54]: This should be another state, not a boolean

File: Assets/Scripts/Digging/Digging.cs [hunk 2]
```diff
@@ -45,13 +47,39 @@ void Update()
...
-            tile.DigTile();
+            if (tile.tileState == TileState.Normal)
+                tile.DigTile();
+            else if (tile.tileState == TileState.Respawning && !rope.InUse && ((rope.gameObject.activeSelf && tile.isRopeIn) || (!rope.gameObject.activeSelf && !tile.isRopeIn)))
+            {
+                playerMovement.MovementDisabled = !playerMovement.MovementDisabled;
+                tile.isRopeIn = !tile.isRopeIn;
+
+                if (tile.isRopeIn)
+
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Digging/Digging.cs:69]: no space

File: Assets/Scripts/Digging/Digging.cs [hunk 3]
```diff
@@ -45,13 +47,39 @@ void Update()
...
+                    StartCoroutine(RopeThrowing(rope.gameObject, tile));
+                else
+                    this.rope.gameObject.SetActive(false);
+                
+            }
+
+        if (this.rope.Saved)
+        {
+
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Digging/Digging.cs:78]: no space

File: Assets/Scripts/Digging/Digging.cs [hunk 4]
```diff
@@ -94,4 +122,18 @@ private Collider GetClosestCollider(RaycastHit[] hits)
 
         return closestCollider;
     }
+
+    public IEnumerator RopeThrowing(GameObject ropeObject, Tile tile)
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Digging/Digging.cs:126]: Name it ThrowRope (as an action)

File: Assets/Scripts/Digging/Digging.cs [hunk 5]
```diff
@@ -94,4 +122,18 @@ private Collider GetClosestCollider(RaycastHit[] hits)
...
+    {
+        Vector3 tileSurfacePosition = new Vector3(tile.transform.position.x, this.transform.position.y, tile.transform.position.z);
+        while (Vector3.Distance(this.transform.position, tileSurfacePosition) > 1.5f)
+        {
+            playerMovement.MoveTowards(ropeObject.transform.forward, 6f);
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Digging/Digging.cs:131]: what's the 6f magic number?

File: Assets/Scripts/Digging/Digging.cs [hunk 6]
```diff
@@ -94,4 +122,18 @@ private Collider GetClosestCollider(RaycastHit[] hits)
 
         return closestCollider;
     }
+
+    public IEnumerator RopeThrowing(GameObject ropeObject, Tile tile)
+    {
+        Vector3 tileSurfacePosition = new Vector3(tile.transform.position.x, this.transform.position.y, tile.transform.position.z);
+        while (Vector3.Distance(this.transform.position, tileSurfacePosition) > 1.5f)
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Digging/Digging.cs:129]: Hardcoded 1.5f, make it an adjustable value

File: Assets/Scripts/Player/PlayerMovement.cs [hunk 7]
```diff
@@ -22,9 +23,11 @@ public class PlayerMovement : MonoBehaviour
     private Vector3 cameraForward;
     private Vector3 cameraRight;
     private int layerMask;
+    public bool MovementDisabled;
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/PlayerMovement.cs:26]: booleans with IsMovementDisabled

File: Assets/Scripts/Player/PlayerMovement.cs [hunk 8]
```diff
@@ -22,9 +23,11 @@ public class PlayerMovement : MonoBehaviour
...
     // Falling vars
     [SerializeField] private float fallingVelocity = 10f;
     private bool isFalling;
+    [SerializeField] private Image img;
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/PlayerMovement.cs:30]: What's this for?

File: Assets/Scripts/Player/PlayerMovement.cs (hunk 7 above)
Comment (danielwiktorczyk) [reply Assets/Scripts/Player/PlayerMovement.cs:26]: or MovementIsDisabled

File: Assets/Scripts/Player/PlayerMovement.cs [hunk 9]
```diff
@@ -195,4 +202,39 @@ private Collider GetClosestCollider(Collider[] hitColliders)
...
+        {
+            this.characterController.Move(Vector3.up * Time.deltaTime * this.movementSpeed);
+            yield return null;
+        }
+        yield return new WaitForSeconds(0.2f);
+        this.transform.position = new Vector3(surfacePosition.x, surfacePosition.y+1f, surfacePosition.z);
+        yield return StartCoroutine(fadeOut());
+    }
+
+    public IEnumerator fadeIn()
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/PlayerMovement.cs:220]: PascalCase

File: Assets/Scripts/Player/PlayerMovement.cs [hunk 10]
```diff
@@ -195,4 +202,39 @@ private Collider GetClosestCollider(Collider[] hitColliders)
...
+    {
+        for (float i = 0; i <= 2; i += Time.deltaTime)
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/PlayerMovement.cs:222]: You can rename i to opacity

Comment (danielwiktorczyk) [reply Assets/Scripts/Player/PlayerMovement.cs:222]: Why 2?

File: Assets/Scripts/Player/PlayerMovement.cs [hunk 11]
```diff
@@ -195,4 +202,39 @@ private Collider GetClosestCollider(Collider[] hitColliders)
...
+        {
+            // set color with i as alpha
+            this.img.color = new Color(0, 0, 0, i);
+            yield return null;
+        }
+    }
+
+    public IEnumerator fadeOut()
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/PlayerMovement.cs:230]: Pascal Case

File: Assets/Scripts/Player/Rope.cs [hunk 12]
```diff
@@ -0,0 +1,82 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class Rope : MonoBehaviour
+{
+    private GameObject highlightLadder;
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/Rope.cs:7]: highlightedLadder

File: Assets/Scripts/Player/Rope.cs [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    private InputManager inputManager;
+    private bool inZone;
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/Rope.cs:9]: isInZone? not sure what this does just yet

File: Assets/Scripts/Player/Rope.cs [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    public bool InUse;
+    public bool Saved;
+    private PlayerMovement playerMovement;
+    private float heightToClimb = 8;
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/Rope.cs:13]: make this editable in inspector

File: Assets/Scripts/Player/Rope.cs [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+
+public class Rope : MonoBehaviour
+{
+    private GameObject highlightLadder;
+    private InputManager inputManager;
+    private bool inZone;
+    public bool InUse;
+    public bool Saved;
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/Rope.cs:11]: IsSaved? 
You should create an enum for all the different states of the Rope, called RopeState

```
public enum RopeState {
   Normal, 
   ...
   }
   ```

File: Assets/Scripts/Tiling/Tile.cs [hunk 16]
```diff
@@ -104,24 +109,28 @@ private IEnumerator PlayBreakingAnimation()
...
     {
+        this.gameObject.layer = LayerMask.NameToLayer("TileMovementCollider");
         transform.position = new Vector3(transform.position.x, 0, transform.position.z);
-
         this.meshRenderer.material = normalMaterial;
-
         this.tileState = TileState.Normal;
     }
 
     public IEnumerator HighlightTileForOneFrame()
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Tiling/Tile.cs:130]: I would leave this code here unchanged. Make a new method for highlighting it for the rope separately

Make sure my scene works for combos!!!

File: Assets/Scripts/Digging/Digging.cs [hunk 17]
```diff
@@ -45,13 +47,39 @@ void Update()
...
+            this.rope.gameObject.SetActive(false);
+            tile.isRopeIn = false;
             return;
+        }
 
         StartCoroutine(tile.HighlightTileForOneFrame());
 
         if (inputManager.GetDigging())
-            tile.DigTile();
+            if (tile.tileState == TileState.Normal)
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Digging/Digging.cs:61]: combine this check in the prior boolean check. Too nested rn

Comment (danielwiktorczyk) [reply Assets/Scripts/Digging/Digging.cs:61]: or refactor this method

File: Assets/Scripts/Digging/Digging.cs [hunk 18]
```diff
@@ -45,13 +47,39 @@ void Update()
...
+                tile.DigTile();
+            else if (tile.tileState == TileState.Respawning && !rope.InUse && ((rope.gameObject.activeSelf && tile.isRopeIn) || (!rope.gameObject.activeSelf && !tile.isRopeIn)))
+            {
+                playerMovement.MovementDisabled = !playerMovement.MovementDisabled;
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Digging/Digging.cs:65]: woah why are we inverting it here? I think you want to do one or the other; either set them true, or false.

**Review by danielwiktorczyk — CHANGES_REQUESTED** [review]

**Review by eli-lam — CHANGES_REQUESTED** [review]

File: Assets/Scripts/Player/Rope.cs [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+    private PlayerMovement playerMovement;
+    private float heightToClimb = 8;
+    private Teammate team;
+    // Start is called before the first frame update
+    void Awake()
+    {
+        this.heightToClimb = 8;
+        this.highlightLadder = this.transform.GetChild(2).gameObject;
```
[the remark this answers is not in the document]
Comment (danielwiktorczyk) [reply Assets/Scripts/Player/Rope.cs:19]: true, this should be a drag and drop from the inspector if this is a prefab

File: Assets/Scripts/Digging/Digging.cs (hunk 18 above)
Comment (zahranikbakht) [reply Assets/Scripts/Digging/Digging.cs:65]: You can throw/unthrow rope with the same button, so each button press results in inverting the movement freeze.

File: Assets/Scripts/Player/Rope.cs (hunk 13 above)
Comment (zahranikbakht) [reply Assets/Scripts/Player/Rope.cs:9]: checks if player is close enough to the ladder to highlight it.

File: Assets/Scripts/Player/PlayerMovement.cs (hunk 8 above)
Comment (zahranikbakht) [reply Assets/Scripts/Player/PlayerMovement.cs:30]: it's just a plain black image used for fade in/fade out

File: Assets/Scripts/Digging/Digging.cs (hunk 5 above)
Comment (zahranikbakht) [reply Assets/Scripts/Digging/Digging.cs:131]: it's just the speed that the character moves toward the rope when throwing one

Comment (zahranikbakht) [reply Assets/Scripts/Digging/Digging.cs:131]: made it adjustable tho it's not really important

File: Assets/Scripts/Player/PlayerMovement.cs (hunk 10 above)
Comment (zahranikbakht) [reply Assets/Scripts/Player/PlayerMovement.cs:222]: made the time adjustable

File: Assets/Scripts/Tiling/Tile.cs (hunk 16 above)
Comment (zahranikbakht) [reply Assets/Scripts/Tiling/Tile.cs:128]: I only added an extra condition for rope throwing. It works in your scene too.

File: Assets/Scripts/Digging/Digging.cs [hunk 20]
```diff
@@ -6,28 +6,31 @@
...
     void Start() => inputManager = InputManager.GetInstance();
 
     void Update()
     {
-        if (playerMovement.IsFalling())
-            return;
-
 
+        if (this.playerMovement.isInUnderground)
+        {
```
Comment (eli-lam) [inline Assets/Scripts/Digging/Digging.cs:31]: Don't need the `{ }`

File: Assets/Scripts/Digging/Digging.cs [hunk 21]
```diff
@@ -94,4 +110,45 @@ private Collider GetClosestCollider(RaycastHit[] hits)
...
         return closestCollider;
     }
+
+    public IEnumerator ThrowRope(GameObject ropeObject, Tile tile)
+    {
+        Vector3 tileSurfacePosition = new Vector3(tile.transform.position.x, this.transform.position.y, tile.transform.position.z);
+        while (Vector3.Distance(this.transform.position, tileSurfacePosition) > maxDistanceToRope)
+        {
+            playerMovement.MoveTowards(ropeObject.transform.forward, speedTowardsRope);
+            yield return 0;
```
Comment (eli-lam) [inline Assets/Scripts/Digging/Digging.cs:120]: Do `yield return null` instead, `yield return 0` allocates memory

File: Assets/Scripts/Player/PlayerMovement.cs [hunk 22]
```diff
@@ -96,13 +103,18 @@ private bool HasFallenInHole()
         }
         else if (hitColliders.Length == 1)
         {
```
Comment (eli-lam) [inline Assets/Scripts/Player/PlayerMovement.cs:105]: Don't need the `{ }`

File: Assets/Scripts/Player/Rope.cs [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+        this.playerMovement.isMovementDisabled = true;
+        var directionToRope = (this.transform.position - this.playerMovement.transform.position).normalized;
+        directionToRope = new Vector3(directionToRope.x, 0,directionToRope.z);
+        var ropePositionWithoutY = new Vector3(this.transform.position.x, this.playerMovement.transform.position.y, this.transform.position.z);
+        this.playerMovement.transform.LookAt(ropePositionWithoutY);
+
+        while (Vector3.Distance(this.playerMovement.transform.position, ropePositionWithoutY) > 1.0f)
+        {
+            this.playerMovement.MoveTowards(directionToRope, 120f);
+            yield return 0;
```
Comment (eli-lam) [inline Assets/Scripts/Player/Rope.cs:81]: `yield return null`

File: Assets/Scripts/Tiling/Tile.cs [hunk 24]
```diff
@@ -131,11 +138,19 @@ private void ChangeMaterialAccordingToCurrentState()
     {
         if (this.isHighlighted)
         {
-            this.meshRenderer.material = highlightedMaterial;
+            if (this.tileState == TileState.Respawning)
+            {
```
Comment (eli-lam) [inline Assets/Scripts/Tiling/Tile.cs:142]: No need for `{ }`

**Review by eli-lam — APPROVED** [review]

File: Assets/Scripts/Digging/Digging.cs [hunk 25]
```diff
@@ -94,4 +109,45 @@ private Collider GetClosestCollider(RaycastHit[] hits)
...
+            playerMovement.MoveTowards(ropeObject.transform.forward, speedTowardsRope);
+            yield return null;
+        }
+        playerMovement.MoveTowards(Vector3.zero, 0);
+        ropeObject.transform.position = new Vector3(tile.transform.position.x, ropeObject.transform.position.y, tile.transform.position.z);
+        ropeObject.SetActive(true);
+        yield return null;
+    }
+
+    public void InterctWithTile(Tile tile)
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Digging/Digging.cs:127]: Interact (typo)

Comment (danielwiktorczyk) [reply Assets/Scripts/Digging/Digging.cs:127]: Why is it called this anyway? This is pretty general. How are we interacting with it?

File: Assets/Scripts/Digging/Digging.cs [hunk 26]
```diff
@@ -94,4 +109,45 @@ private Collider GetClosestCollider(RaycastHit[] hits)
...
+    {
+        
+        if (tile.tileState == TileState.Normal)
+            tile.DigTile();
+        else if (rope.ropeState == RopeState.Normal && ((this.rope.gameObject.activeSelf && tile.tileState == TileState.RopeIn) || (!this.rope.gameObject.activeSelf && tile.tileState == TileState.Respawning)))
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Digging/Digging.cs:132]: make this multiple lines. Currently, this compound condition is hard to follow

File: Assets/Scripts/Player/PlayerMovement.cs [hunk 27]
```diff
@@ -16,14 +17,16 @@ public class PlayerMovement : MonoBehaviour
...
     [SerializeField] private float fallingVelocity = 10f;
     private bool isFalling;
-
+    public bool isInUnderground;
+    [SerializeField] private Image img;
+    [SerializeField] private float timeToFadeIn = 2f;
     void Awake()
     {
-        layerMask = LayerMask.GetMask("TileMovementCollider");
+        layerMask = LayerMask.GetMask("TileMovementCollider") | LayerMask.GetMask("Underground");
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/PlayerMovement.cs:29]: Rename this `layerMask`. The value you are setting it to is not a general one

File: Assets/Scripts/Player/PlayerMovement.cs [hunk 28]
```diff
@@ -94,14 +101,16 @@ private bool HasFallenInHole()
...
-        }
+        else if (hitColliders.Length == 1)   
+            closestCollider = hitColliders[0];   
+
         else if (hitColliders.Length > 1)
-        {
             closestCollider = GetClosestCollider(hitColliders);
-        }
+
+        if (closestCollider.gameObject.layer == LayerMask.NameToLayer("Underground"))
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/PlayerMovement.cs:110]: This doesn't seem like it will work all the time. 

It would be better to dynamically get the 1 normal plane, the 1 underground plane, and keep those references. Then, when you need to check this, you just do Vector3.Distance and find which one we're closest to. It's also overkill to do a collision check every frame for this!

File: Assets/Scripts/Player/PlayerMovement.cs [hunk 29]
```diff
@@ -133,4 +142,39 @@ private Collider GetClosestCollider(Collider[] hitColliders)
     }
 
     public bool IsFalling() => this.isFalling;
+    public void MoveTowards(Vector3 direction, float speed) => this.characterController.Move(direction * Time.deltaTime * speed);
+
+    public IEnumerator ClimbRope(float height,Vector3 surfacePosition)
+    {
+        StartCoroutine(FadeIn());
+        while (this.transform.position.y < -70 + height)
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/PlayerMovement.cs:150]: the -70 is a magic number

File: Assets/Scripts/Player/PlayerMovement.cs [hunk 30]
```diff
@@ -94,14 +101,16 @@ private bool HasFallenInHole()
...
-        {
             closestCollider = GetClosestCollider(hitColliders);
-        }
+
+        if (closestCollider.gameObject.layer == LayerMask.NameToLayer("Underground"))
+            isInUnderground = true;
+        else
+            isInUnderground = false;
 
         if (this.transform.position.y < -20)
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/PlayerMovement.cs:115]: this -20 is a magic number

File: Assets/Scripts/Player/PlayerMovement.cs [hunk 31]
```diff
@@ -133,4 +142,39 @@ private Collider GetClosestCollider(Collider[] hitColliders)
...
+    public IEnumerator ClimbRope(float height,Vector3 surfacePosition)
+    {
+        StartCoroutine(FadeIn());
+        while (this.transform.position.y < -70 + height)
+        {
+            this.characterController.Move(Vector3.up * Time.deltaTime * this.movementSpeed);
+            yield return null;
+        }
+        yield return new WaitForSeconds(0.2f);
+        this.transform.position = new Vector3(surfacePosition.x, surfacePosition.y+1f, surfacePosition.z);
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Player/PlayerMovement.cs:156]: Can do `this.transform.position += Vector3.up * 1f`
Also, that 1f is a magic number

File: Assets/Scripts/Tiling/Tile.cs [hunk 32]
```diff
@@ -76,13 +81,15 @@ private void UnstableUpdate()
...
         if (progress <= 0)
             Break();
+    
     }
 
     private void Break()
     {
-        StartCoroutine(PlayBreakingAnimation());
-
+        //StartCoroutine(PlayBreakingAnimation());
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Tiling/Tile.cs:89]: why are you removing this animation??

File: Assets/Scripts/Tiling/Tile.cs [hunk 33]
```diff
@@ -76,13 +81,15 @@ private void UnstableUpdate()
...
+        this.gameObject.layer = LayerMask.NameToLayer("TileRope");
```
Comment (danielwiktorczyk) [inline Assets/Scripts/Tiling/Tile.cs:90]: You shouldn't change the layer dynamically. Let's VC and I can help you with this

File: Assets/Scripts/Tiling/Tile.cs (hunk 16 above)
Comment (danielwiktorczyk) [reply Assets/Scripts/Tiling/Tile.cs:128]: But there are now multiple types of highlighting: simple dig, combo dig, and rope. We are adding too many conditions to this, and should instead make seperate methods that we call based on the context. E.g. HighlightWithRope, HighlightForCombo, HighlightForSimpleDig

**Review by danielwiktorczyk — CHANGES_REQUESTED** [review]
There are still some things I'd like to go over with you. Let's jump on a call!

**Review by danielwiktorczyk — APPROVED** [review]

**Comment by danielwiktorczyk** [comment]
Excellent work!! Love the hole system. Merging-in-the-holeeeee!!
